File: typescript/stack/stack.ts

class StackNode<T> {
    constructor(public value: T, public next: StackNode<T> | null = null) { }   
}

export class Stack<T> {
    private head: StackNode<T> | null = null
    private length: number = 0

    public push(value: T) {
        const node = new StackNode(value, this.head)
        this.head = node
        this.length++
    }

    public pop() {
        if (!this.head) return
        const node = this.head
        this.head = node.next
        this.length--
        return node.value
    }
    public top() {
        return this.head?.value
    }
    public isEmpty() {
        return this.length === 0
    }

    public size() {
        return this.length
    }
}

File: go/stack/stack.go
package stack

type StackNode struct {
	Data int
	Next *StackNode
}

type Stack struct {
	head   *StackNode
	length int
}

func New() *Stack {
	return &Stack{
		head:   nil,
		length: 0,
	}
}

func (s *Stack) Push(data int) {
	node := &StackNode{
		Data: data,
		Next: s.head,
	}
	s.head = node
	s.length++
}

func (s *Stack) Pop() int {
	if s.head == nil {
		return -1
	}
	data := s.head.Data
	s.head = s.head.Next
	s.length--
	return data
}

func (s *Stack) Top() int {
	if s.head == nil {
		return -1
	}
	return s.head.Data
}

func (s *Stack) Size() int {
	return s.length
}

func (s *Stack) IsEmpty() bool {
	return s.length == 0
}
